cfn-resources/integration-tests/utility/util.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2023 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utility
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"os/exec"
+	"testing"
+)
+
+func GetRandNum() *big.Int {
+	randInt, _ := rand.Int(rand.Reader, big.NewInt(10000))
+	return randInt
+}
+
+func FailNowIfError(t *testing.T, msgf string, err error) {
+	if err != nil {
+		t.Errorf(msgf, err.Error())
+		t.FailNow()
+	}
+}
+
+func RunCleanupScript(t *testing.T, rctx ResourceContext) {
+	output, err := runShScript("../cleanup_cfn.sh")
+	FailNowIfError(t, fmt.Sprintf("Error when executing cleanup script. Output: %s\n", output)+"%v", err)
+
+	t.Logf("E2E test resource type %s successfully de-registered from private registry!\n", rctx.ResourceTypeNameForE2e)
+}
+
+func PublishToPrivateRegistry(t *testing.T, rctx ResourceContext) {
+	t.Setenv("RESOURCE_TYPE_NAME", rctx.ResourceTypeName)
+	t.Setenv("RESOURCE_TYPE_NAME_FOR_E2E", rctx.ResourceTypeNameForE2e)
+	t.Setenv("E2E_RAND_SUFFIX", rctx.E2eRandSuffix)
+	t.Setenv("RESOURCE_DIRECTORY_NAME", rctx.ResourceDirectory)
+
+	output, err := runShScript("../publish_cfn_to_registry.sh")
+	FailNowIfError(t, fmt.Sprintf("Error when executing publishing script. Output: %s\n", output)+"%v", err)
+
+	t.Logf("New E2E test resource type %s successfully published to private registry!\n", rctx.ResourceTypeNameForE2e)
+
+	t.Cleanup(func() {
+		RunCleanupScript(t, rctx)
+	})
+}
+
```

Comment:
Can't we assume that the resource has been already released using the Makefile before the test is run? We can fail the test if the ENV VAR  `RESOUCE_TYPE_NAME` is not available

Reply:
My thought process was basically to have the _test.go to be the driver of everything and the only point of entry. Are there any specific benefits of doing this using Makefile?

Reply:
> My thought process was basically to have the _test.go to be the driver of everything and the only point of entry.

right now `publish_cfn_to_registry.sh` is publishing and publishing the resource in the same script. I don't understand how this logic can work since the script will unpublish the resource before you will be able to use it in a stack. 
This is way I asked you to share the log of a run of the test so I can understand the correct flow and if I am missing something.

> Also how will cleanup be properly addressed with Makefile?

- You can split `publish_cfn_to_registry.sh` in two scripts:`publish_cfn_to_registry.sh` and `clean-up.sh`, 
- create two separate Makefile commands.  
- In the GitHub action you call first `make e2e-test-init # this calls publish_cfn_to_registry.sh`, you then run your test `go test ..`, finally you call `make e2e-clean # this calls clean-up.sh`

> Are there any specific benefits of doing this using Makefile?

yes, less code to maintain. In general, you don't use go to run script files if you can avoid it.

Reply:
> right now publish_cfn_to_registry.sh is publishing and publishing the resource in the same script.

This is not true. You have split the script in two separate scripts 👍